scRNA-seq-advanced/setup/rms/integrate_rms.R

```diff
@@ -0,0 +1,131 @@
+
+# create list of SCE objects 
+if(is.null(opt$input_file_list)){
+  stop("List of input files containing individual SCE objects to merge is missing.")
+} else {
+  sce_paths <- unlist(stringr::str_split(opt$input_file_list, ' '))
+} 
+
+# check that at least 2 SCE files have been provided 
+if(length(sce_paths) < 2){
+  stop("Must provide at least 2 SCE files to integrate.")
+}
+
+# extract sample ids from file names 
+sample_ids <- stringr::str_remove(basename(sce_paths), ".rds")
+
+# read in list of SCE objects 
+sce_list <- purrr::map(
+  sce_paths,
+  readr::read_rds
+)
+# name list with sample ids 
+names(sce_list) <- sample_ids
+
+
+# Make output directory if it doesn't exist
+output_dir <- dirname(opt$integrated_sce_file)
+if (!(dir.exists(output_dir))) {
+  dir.create(output_dir)
+}
+
+# Merge SCE objects ------------------------------------------------------------
+
+# set up function to re-format individual SCEs prior to merging
+format_sce_list <- function(sce, sce_name, shared_genes) {
+  
+  # Reduce to shared genes
+  sce <- sce[shared_genes,]
+  
+  # Update the colData row names so we can know where cells came from back
+  rownames(colData(sce)) <- glue::glue("{sce_name}-{rownames(colData(sce))}")
+  
+  # Update the rowData names, except `gene_symbol`
+  names(rowData(sce)) <- ifelse(names(rowData(sce)) == "gene_symbol",
+                                "gene_symbol",
+                                glue::glue("{sce_name}-{rownames(rowData(sce))}")
+                                
+  )
+  # Add in sample-level information as stand-alone column
+  colData(sce)$sample <- sce_name
+  
+  # Return
+  return(sce)
+}
+
+# grab shared genes 
+shared_genes <- sce_list %>%
+  # get rownames for each entry in sce_list
+  purrr::map(rownames) %>%
+  # find their intersection
+  purrr::reduce(intersect)
+
+
+# Prepare the sce_list for combining
+sce_list <- purrr::imap(sce_list,
+                        format_sce_list,
+                        shared_genes = shared_genes)
+
+# combine into one sce
+combined_sce <- do.call(cbind, sce_list)
+
+# Dimension Reduction ----------------------------------------------------------
+
+# select combined highly variable genes
+gene_variance <- scran::modelGeneVar(combined_sce,
+                                     # For combined SCEs, specify the group column groups:
+                                     block = combined_sce$sample)
+hvg_list <- scran::getTopHVGs(gene_variance,
+                              n = 2000)
+
+# Add PCA and UMAP into the SCE
+combined_sce <- scater::runPCA(combined_sce,
+                               subset_row = hvg_list) %>%
+  scater::runUMAP(dimred = "PCA")
+
+# Integration ------------------------------------------------------------------
+
+# perform fastMNN integration 
+integrated_sce <- batchelor::fastMNN(combined_sce,
+                                     batch = combined_sce$sample)
```

Comment:
Do we want to use `subset.row = hvg_list` here? That would better parallel what we have in the instruction notebook, if people were trying to recreate this object. If we do that, we will need to be sure to also include `correct.all = TRUE` to get out the full reconstructed matrix. 

 I do think it is worth adding the reconstructed gene expression  to the output, if only for future-proofing purposes. We _might_ want to look at it.